### PR TITLE
Add eth1 cache api

### PR DESCRIPTION
**Pull request by mehdi-aouadi**
## PR Description

Add a REST API endpoint to list the cached eth1 blocks: `/teku/v1/beacon/pool/eth1cache`

## Fixed Issue(s)

Fixes issue https://github.com/ConsenSys/teku/issues/5671

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/Eth1DataCache.java [hunk 1]
```diff
@@ -124,6 +126,10 @@ public Eth1Data getEth1Vote(BeaconState state) {
     return vote.orElse(defaultVote);
   }
 
+  public List<Eth1Data> getAllEth1Blocks() {
+    return new ArrayList<>(this.eth1ChainCache.values());
```
Comment (tbenr) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/Eth1DataCache.java:130]: I think you can return `Collection<Eth1Data>` here. And keep returning it up to the handler.  SerializableTypeDefinition should handle it

Comment (mehdi-aouadi) [reply beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/Eth1DataCache.java:130]: Done

File: infrastructure/restapi/src/testFixtures/java/tech/pegasys/teku/infrastructure/restapi/StubRestApiRequest.java [hunk 2]
```diff
@@ -152,6 +152,14 @@ public void respondWithCode(final int statusCode) {
     verifyResponseMatchesMetadata();
   }
 
+  @Override
+  public void respondWithCode(int statusCode, Object response) throws JsonProcessingException {
+    assertThat(this.responseCode).isEqualTo(CODE_NOT_SET);
+    this.responseCode = statusCode;
+    this.responseObject = response;
```
Comment (rolfyone) [inline infrastructure/restapi/src/testFixtures/java/tech/pegasys/teku/infrastructure/restapi/StubRestApiRequest.java:159]: should make sure we're not setting an object that has been set as well...

File: infrastructure/restapi/src/testFixtures/java/tech/pegasys/teku/infrastructure/restapi/StubRestApiRequest.java [hunk 3]
```diff
@@ -152,6 +152,14 @@ public void respondWithCode(final int statusCode) {
     verifyResponseMatchesMetadata();
   }
 
+  @Override
+  public void respondWithCode(int statusCode, Object response) throws JsonProcessingException {
```
Comment (rolfyone) [inline infrastructure/restapi/src/testFixtures/java/tech/pegasys/teku/infrastructure/restapi/StubRestApiRequest.java]: respondWithCode was to indicate we're not setting a body. i think you're wanting to use `respondError` and just pass in the error code and string message that goes into the `HttpErrorResponse`... will comment above to make it more obvious, but basically we don't want this method `respondWithCode(int,object)` to exist...

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/tekuv1/beacon/GetEth1DataCache.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    adapt(ctx);
+  }
+
+  @Override
+  public void handleRequest(RestApiRequest request) throws JsonProcessingException {
+    request.header(Header.CACHE_CONTROL, CACHE_NONE);
+    Collection<Eth1Data> eth1CachedBlocks = this.eth1DataProvider.getEth1CachedBlocks();
+    if (eth1CachedBlocks.isEmpty()) {
+      request.respondWithCode(
+          SC_NOT_FOUND, new HttpErrorResponse(SC_NOT_FOUND, "Eth1 blocks cache is empty"));
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/tekuv1/beacon/GetEth1DataCache.java]: commented in the new function, but basically this is just `request.respondError(SC_NOT_FOUND, "Eth1 block cache is empty")`

File: infrastructure/restapi/src/main/java/tech/pegasys/teku/infrastructure/restapi/endpoints/JavalinRestApiRequest.java [hunk 5]
```diff
@@ -145,6 +145,11 @@ public void respondWithUndocumentedCode(final int statusCode) {
     context.status(statusCode);
   }
 
+  @Override
+  public void respondWithCode(int statusCode, Object response) throws JsonProcessingException {
+    respond(statusCode, response, getResponseOutputStream());
+  }
+
```
Comment (rolfyone) [inline infrastructure/restapi/src/main/java/tech/pegasys/teku/infrastructure/restapi/endpoints/JavalinRestApiRequest.java]: will disappear in line with other comments

File: infrastructure/restapi/src/main/java/tech/pegasys/teku/infrastructure/restapi/endpoints/RestApiRequest.java [hunk 6]
```diff
@@ -34,6 +34,8 @@ public interface RestApiRequest {
 
   void respondWithCode(int statusCode);
 
+  void respondWithCode(int statusCode, Object response) throws JsonProcessingException;
+
```
Comment (rolfyone) [inline infrastructure/restapi/src/main/java/tech/pegasys/teku/infrastructure/restapi/endpoints/RestApiRequest.java]: again, will disappear

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/tekuv1/beacon/GetEth1DataCache.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+              Function.identity())
+          .build();
+
+  private final Eth1DataProvider eth1DataProvider;
+
+  public GetEth1DataCache(Eth1DataProvider eth1DataProvider) {
+    super(
+        EndpointMetadata.get(ROUTE)
+            .operationId("getTekuV1BeaconPoolEth1cache")
+            .summary("Get all cached Eth1 blocks")
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/tekuv1/beacon/GetEth1DataCache.java]: this is the title in swagger, and tends to only have a small amount of room on teku swagger site: https://consensys.github.io/teku/#stable
You'll see some under teku aren't great... but anyway I think I would set the summary as `Get cached eth1 blocks`, and the description to something a little more descriptive, like `Get all of the eth1 blocks currently cached by the beacon node, that could be considered for inclusion during block production.`

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/tekuv1/beacon/GetEth1DataCache.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            .withNotFoundResponse()
+            .build());
+    this.eth1DataProvider = eth1DataProvider;
+  }
+
+  @OpenApi(
+      path = ROUTE,
+      method = HttpMethod.GET,
+      summary = "Get all cached Eth1 blocks",
+      description = "Get all cached Eth1 blocks.",
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/tekuv1/beacon/GetEth1DataCache.java]: keeping these consistent with above is a good idea, same comment applies :)

**Comment by rolfyone** [comment]
Looks pretty good, some minor considerations added :)

File: infrastructure/restapi/src/testFixtures/java/tech/pegasys/teku/infrastructure/restapi/StubRestApiRequest.java (hunk 3 above)
Comment (mehdi-aouadi) [reply infrastructure/restapi/src/testFixtures/java/tech/pegasys/teku/infrastructure/restapi/StubRestApiRequest.java]: My bad, missed the `respondError`

File: data/beaconrestapi/src/integration-test/resources/tech/pegasys/teku/beaconrestapi/beacon/paths/_teku_v1_beacon_pool_eth1cache.json [hunk 9]
```diff
@@ -0,0 +1,28 @@
+{
+  "get": {
+    "tags": [
+      "Teku"
+    ],
+    "summary": "Get all cached Eth1 blocks",
+    "description": "Get all cached Eth1 blocks.",
```
Comment (tbenr) [inline data/beaconrestapi/src/integration-test/resources/tech/pegasys/teku/beaconrestapi/beacon/paths/_teku_v1_beacon_pool_eth1cache.json]: You'll need to update this too

**Review by tbenr — APPROVED** [review]
LGTM! Nice first PR!
Just need to update the API summary in the integration tests check

**Pull request merged**